### DISCUSSION TITLE
Docs for using with Tamagui

### DISCUSSION THREAD

**Discussion by ryan0x44** [vikejs/vike#981] [Discussion]
### Description

I'm not even sure if vite-plugin-ssr can work with Tamagui but it would be awesome if it could!

Tamagui definitely works with Vite https://tamagui.dev/docs/guides/vite
I've got an example of it here https://github.com/ryan0x44/tamagui-vite-ssr/tree/tamagui-vite-no-ssr

My attempts to get vite-plugin-ssr working with Tamagui are here on the main branch: https://github.com/ryan0x44/tamagui-vite-ssr - however I get errors when trying various combinations of this.

It would be a huge win for vite-plugin-ssr over Next.js if it fully supported SSR with Tamagui!

**Comment by brillout** [comment]
I agree. Contribution welcome to create a working example.

> I get errors

Feel free to let me know about these.

**Reply by kzlar** [reply]
Hey @brillout , I'm also having issues with this.
Here are the steps to reproduce:
1. Clone or use Github Codespace to get @ryan0x44 's code - https://github.com/ryan0x44/tamagui-vite-ssr
2. `yarn && yarn dev` then open browser. Error that I get:
```
$ TAMAGUI_TARGET=web vite
4:37:27 AM [vite-plugin-ssr][Warning] We recommend setting /workspaces/tamagui-vite-ssr/package.json#type to "module" (and therefore writing ESM code instead of CJS code), see https://vite-plugin-ssr.com/CJS
  VITE v4.5.0  ready in 1647 ms
  ➜  Local:   http://localhost:3000/
  ➜  Network: use --host to expose
  ➜  press h to show help
4:37:38 AM [vite-plugin-ssr][request(1)] HTTP request: /
(node:2818) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
4:37:39 AM [vite][request(1)] Error when evaluating SSR module /node_modules/@tamagui/web/dist/esm/config.js: failed to import "@tamagui/constants"
|- /workspaces/tamagui-vite-ssr/node_modules/@tamagui/constants/dist/esm/index.js:1
export * from "./constants";
^^^^^^

SyntaxError: Unexpected token 'export'
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1153:20)
    at Module._compile (node:internal/modules/cjs/loader:1205:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Function.Module._load (node:internal/modules/cjs/loader:938:12)
    at cjsLoader (node:internal/modules/esm/translators:284:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:234:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:217:25)
    at ModuleLoader.import (node:internal/modules/esm/loader:316:24)
┌─────────────────────────────────────────────────────────────────────┐
│ Error isn't helpful? See https://vite-plugin-ssr.com/errors#verbose │
└─────────────────────────────────────────────────────────────────────┘
4:37:39 AM [vite-plugin-ssr][request(1)] HTTP response / ERR
Sourcemap for "/workspaces/tamagui-vite-ssr/node_modules/@tamagui/font-inter/dist/esm/index.js" points to missing source files
```

Admittedly I'm a complete newbie about the whole CJS/ESM saga but it sounds like that's what's going on here? Note that Tamagui doesn't work if you set `"type":"module"` according to the [docs](https://tamagui.dev/docs/guides/vite) and the [maintainer](https://discord.com/channels/909986013848412191/909986013848412194/1165055082761551913).
Would you happen to have any pointers into how to solve this?

Thank you 🙏

**Reply by brillout** [reply]
It's unrelated to Vike, AFAICT it's a tamagui issue CC @natew.

**Reply by natew** [reply]
We don’t set type module because it simply broke too many things. React Native doesn’t really support it well enough, and last I remember one of the big issues is that if you do file specific imports (with the .js ending) RN won’t pick up the platform imports anymore (.native.js), amongst other issues. I tried pretty hard to make it work last time but ended up bailing since vite works just fine without it. 

Seems like vite plugin ssr shouldn’t enforce type module if Vite doesn’t?

**Reply by brillout** [reply]
Vike (vite-plugin-ssr has been renamed Vike) doesn't enforce `type: "module"`. (In general, Vike is meticulously designed to not break things.) Vike merely *recommends* users to set `type: "module"` on their *own* `package.json`; nothing is being enforced.

From the error I'm seeing, it does seem like a Tamagui/Vite issue.

**Reply by natew** [reply]
The Tamagui vite starter works just fine, as do many other apps. Not sure what’s going on in this specific one but it’s likely specific to this app setup then.

**Reply by redbar0n** [reply]
So if the developer follows Vike's recommended setup they'd potentially get this issue.

The issue could be a problem between ESM and Metro / Expo EAS ?
If so, then this might be a devuser-land workaround for some:
https://github.com/expo/eas-cli/issues/1854

MUI is experiencing a similar issue:
https://github.com/mui/material-ui/issues/37335

So it may be a problem in Vite itself, which was attempted fixed:
https://github.com/vitejs/vite/pull/13370

This may be a workaround for Vike:
https://github.com/vitejs/vite/pull/8729

**Comment by kzlar** [comment]
Yeah @ryan0x44 's tamagui-vite-no-ssr branch (https://github.com/ryan0x44/tamagui-vite-ssr/tree/tamagui-vite-no-ssr) does work.
Here's the diff https://github.com/ryan0x44/tamagui-vite-ssr/compare/tamagui-vite-no-ssr...main

Not really sure what's missing here in terms of setup. Perhaps something is conflicting between vike and tamagui plugins?

**Comment by kzlar** [comment]
Hey all, anyone happen to know if this is possible now?

**Reply by natew** [reply]
The tamagui Vite plugin in general is a lot more capable now, and tamagui core should “just work”. Do you see any errors?

you can also try npm create vxrn for a beta full stack Vite thing.
